stdlib/src/utils/string_slice.mojo

```diff
@@ -26,6 +26,129 @@ alias StaticString = StringSlice[ImmutableStaticLifetime]
+    U+D000..U+D7FF     | ED         | 80..***9F***| 80..BF     |             |
+    U+E000..U+FFFF     | EE..EF     | 80..BF      | 80..BF     |             |
+    U+10000..U+3FFFF   | F0         | ***90***..BF| 80..BF     | 80..BF      |
+    U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+    U+100000..U+10FFFF | F4         | 80..***8F***| 80..BF     | 80..BF      |
+    .
+    """
+
+    constrained[width >= 4, "width must be >= 4."]()
+    var ptr = DTypePointer(data)
+    var iter_len = length
+    # TODO: implement a faster algorithm like https://github.com/cyb70289/utf8
+    # and benchmark the difference.
+
+    fn invalid_special(b0: UInt8, b1: UInt8) -> Bool:
+        """Special unicode ranges."""
+        if b0 == 0xE0 and not (UInt8(0xA0) <= b1 <= UInt8(0xBF)):
+            return True
+        elif b0 == 0xED and not (UInt8(0x80) <= b1 <= UInt8(0x9F)):
+            return True
+        elif b0 == 0xF0 and not (UInt8(0x90) <= b1 <= UInt8(0xBF)):
+            return True
+        elif b0 == 0xF4 and not (UInt8(0x80) <= b1 <= UInt8(0x8F)):
+            return True
+        return False
+
+    var idx = 0
+    while iter_len >= width:
+        var d = ptr.offset(idx).simd_strided_load[width](1)
+        var comp = d < 0b1000_0000
+        if comp.reduce_and():
+            idx += width
+            iter_len -= width
+            continue
+        var byte_types = countl_zero(~(d & UInt8(0b1111_0000)))
+        var first_byte_type = byte_types[0]
+        if first_byte_type == 0:
+            for i in range(1, width):
+                if byte_types[i] != 0:
+                    idx += i
+                    iter_len -= i
+                    break
+            continue
+
+        alias vec_t = SIMD[DType.uint8, 4]
+        alias n4 = vec_t(4, 1, 1, 1)
+        alias n3 = vec_t(3, 1, 1, 0)
+        alias n3_m = vec_t(0b111, 0b111, 0b111, 0)
+        alias n2 = vec_t(2, 1, 0, 0)
+        alias n2_m = vec_t(0b111, 0b111, 0, 0)
+        var vec = byte_types.slice[4]()
+        var valid_n4 = (vec == n4).reduce_and()
+        var valid_n3 = ((vec & n3_m) == n3).reduce_and()
+        var valid_n2 = ((vec & n2_m) == n2).reduce_and()
+        if not (valid_n4 or valid_n3 or valid_n2):
+            return False
+
+        if invalid_special(d[0], d[1]):
+            return False
+        elif first_byte_type == 2 and d[0] < UInt8(0b1100_0010):
+            return False
+        idx += int(first_byte_type)
+        iter_len -= int(first_byte_type)
+
+    @parameter
+    fn invalid[amnt: Int](i: Int) -> Bool:
+        if i + amnt > iter_len:
+            return True
+
+        @parameter
+        for j in range(1, amnt + 1):
+            if countl_zero(~(ptr[i + j] & UInt8(0b1111_0000))) != 1:
+                return True
+        return invalid_special(ptr[i], ptr[i + 1])
+
+    idx = length - iter_len
+    while idx < length:
+        var val = ptr[idx]
+        var byte_type = countl_zero(~(val & UInt8(0b1111_0000)))
```

Comment:
Can you use the `_utf8_byte_type` helper?

Reply:
I still need to use the countl_zero directly since it can operate on SIMD and not only single byte